Emacs cleanup

dudebout opened the pull request:
###### Motivation for this change

Remove old code in the derivation. Make more explicit the fact that some inputs and steps are optional.
###### Things done
- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [x] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

dudebout committed:
emacs: do not use builder file

Do not use the builder file since:
* it was only decribing a generic build
* it only contained some of the phases of the generic build, the other
  ones being in default.nix; this split made understanding the
  derivation more difficult
* having the phases of a generic build in the derivation makes them
  easier to override

Do not remove builder.sh as it is used by macport-24.5.nix.

Before copying the content of builder.sh to default.nix, two chunks were
removed as they were redundant for one, and not relevant any longer for
the other one:
* the preInstall phase was redundant; the substitutions described in it
had already happened in the preConfigure phase
* the linux crt{1,i,n}.o patch was not relevant anymore as all mentions
of these files have been removed in emacs 25.

dudebout committed:
emacs: cosmetic changes

Make gtk2 optional, since it is only used if withGTK2 is not null.

Put gtk2 just above gtk3 in a couple of places for increased
readability.

Replace a single-element list lib.optionals with lib.optional.

dudebout committed:
emacs: remove a redundant NIX_CFLAGS_COMPILE

By default, the NIX_CFLAGS_COMPILE contains the include directory of all
of the packages in buildInputs.

This particular NIX_CFLAGS_COMPILE added the include/cairo
directory. This was properly copy-pasted from a derivation written
before glib.flattenInclude was used to copy the content of include/cairo
into include.

dudebout committed:
emacs: create nativeBuildInputs

Split the buildInputs list into buildInputs and nativeBuildInputs.

Make some of the nativeBuildInputs only required if the derivation is
built directly from the repository, as opposed to, from a tarballed
release. Similarly, only run autogen.sh in this case.

Make the cleanup of .elc files only required if the derivation is built
from a local directory. Move that step into preConfigure for symmetry
with the optional autogen.sh run.

the-kenny commented on pkgs/applications/editors/emacs/default.nix:
```diff
@@ -67,11 +61,17 @@ stdenv.mkDerivation rec {
-  preBuild = ''
+  preConfigure = lib.optionalString srcRepo ''
+    ./autogen.sh
+  '' + lib.optionalString srcLocal ''
```
Is it really necessary to control this via that argument to the expression? We could just run it all the time and simplify the expression by removing `srcLocal`

dudebout replied:
I find having steps meant for working from the git repository mixed with the "regular" steps (i.e. meant to be run from a proper tarball) confusing. This makes the purpose of said steps harder to decipher after the fact, and requires doing a lot of git archeology to clean them up.

I think that the srcRepo option makes sense, as you can point to a git rev, set that option and expect the whole compilation to go through.

Regarding the srcLocal one, I can see two courses of action:

1. Getting rid of it: As it currently stands, it provides only a very incomplete set of (one) steps. The C objects files are not cleaned, etc.
2. Keeping it: It would be really nice to have a more robust set of steps allowing you to use this derivation for emacs development. Maybe using something based off of https://github.com/jwiegley/nix-config/blob/master/emacsHEAD.nix

At this stage, I am for removing srcLocal and the associated step, and revisit the issue in the future.

dudebout replied:
I just realized that all the tarballs already contain byte-compiled elisp files. Therefore that step was really detrimental to compilation time for the general case.

the-kenny commented on pkgs/applications/editors/emacs/default.nix:
```diff
@@ -67,11 +61,17 @@ stdenv.mkDerivation rec {
+    substituteInPlace lisp/international/mule-cmds.el \
+      --replace /usr/share/locale ${gettext}/share/locale
+
+    for i in Makefile.in {src,lib-src,leim}/Makefile.in; do
```
Perhaps `find` to search for `Makefile.in`?

dudebout replied:
Very good idea. That will make sure that all the Makefile.in that actually need it are covered. The following ones are currently not patched despite using /bin/pwd:
* ./Makefile.in
* ./nextstep/Makefile.in
* ./nt/Makefile.in

The following ones are patched but are not using /bin/pwd:
* leim/Makefile.in
* src/Makefile.in

dudebout committed:
emacs: remove srcLocal

Cleaning the .elc files is only needed when developing emacs
locally. This does not really belong in the derivation from a tarball.

dudebout committed:
emacs: use find to identify files to patch

The explicit list was not covering all the files requiring
patching (/bin/pwd was still being used), and was covering files which
did not require patching (misleading to the reader).

dudebout commented:
@the-kenny: comments addressed.

DamienCassou commented:
Did we really need the 6 commits?

dudebout commented:
Thanks @the-kenny.

@DamienCassou I had made 6 commits to document each step and help in the review. Each commit had its rationale in the commit message. Anyway, the commits were squashed pre-merge. The problem with that squashing is that all the information in the commit messages is gone from the repo and only lives in github.

the-kenny commented:
@DamienCassou Note that I squashed the commits while merging the PR
